extensions/replication/queueProcessor/QueueProcessor.js

```diff
@@ -221,48 +223,66 @@ class QueueProcessor {
 
     _getData(entry, cb) {
         this.log.debug('getting data', { entry: entry.getLogInfo() });
-        const req = this.S3source.getObject({
-            Bucket: entry.getBucket(),
-            Key: entry.getObjectKey(),
-            VersionId: entry.getEncodedVersionId() });
-        const incomingMsg = req.createReadStream();
-        req.on('error', err => {
-            this.log.error('error response getting data from S3',
-                           { entry: entry.getLogInfo(),
-                             origin: this.sourceConfig.s3,
-                             error: err.message,
-                             httpStatus: err.statusCode });
-            incomingMsg.emit('error', err);
+        const locations = entry.getLocation();
+        return async.timesLimit(locations.length, LIMIT, (i, next) => {
```

Comment:
For performance reasons, and to avoid an unnecessary "barrier" effect, it could be interesting to replace the "get all streams first and put next" where gets likely involve some metadata lookups on the source for each part and take some non-negligible time, by a "get and start putting ASAP" for each individual stream. This would reduce wait times and increase throughput by some amount.

Basically, the async processing callback would englobe the two operations `get` and `put` serialized in a single processing unit for each part (you may use `async.waterfall` but not sure it's worth it for just two calls).